harvester/harvest.py

```diff
@@ -0,0 +1,497 @@
+
+    # making these fields read-only
+    # (these values can be set during initialization though)
+    _title: str
+    _url: str
+    _extract_type: str  # dcatus vs waf
+    _waf_config: dict = field(default_factory=lambda: {})
+    _extra_source_name: str = "harvest_source_name"  # TODO: update this
+    _dataset_schema: dict = field(
+        default_factory=lambda: open_json(
+            ROOT_DIR / "data" / "dcatus" / "schemas" / "dataset.json"
+        )
+    )
+
+    # not read-only because these values are added after initialization
+    # making them a "property" would require a setter which, because they are dicts,
+    # means creating a custom dict class which overloads the __set_item__ method
+    # worth it? not sure...
+    records: dict = field(default_factory=lambda: {})
+    ckan_records: dict = field(default_factory=lambda: {})
+
+    def __post_init__(self) -> None:
+        # putting this in the post init to avoid an
+        # undefined error with the string formatted vars
+        self._ckan_query: dict = {
+            "q": f'{self._extra_source_name}:"{self._title}"',
+            "fl": [
+                f"extras_{self._extra_source_name}",
+                "extras_dcat_metadata",
+                "extras_identifier",
+                # TODO: add last_modified for waf
+            ],
+        }
+
+    @property
+    def title(self) -> str:
+        return self._title
+
+    @property
+    def url(self) -> str:
+        return self._url
+
+    @property
+    def extract_type(self) -> str:
+        return self._extract_type
+
+    @property
+    def waf_config(self) -> dict:
+        return self._waf_config
+
+    @property
+    def extra_source_name(self) -> str:
+        return self._extra_source_name
+
+    @property
+    def dataset_schema(self) -> dict:
+        return self._dataset_schema
+
+    @property
+    def ckan_query(self) -> dict:
+        return self._ckan_query
+
+    def get_title_from_fgdc(self, xml_str: str) -> str:
+        tree = ET.ElementTree(ET.fromstring(xml_str))
+        return tree.find(".//title").text
+
+    def harvest_to_id_hash(self, records: list[dict]) -> None:
+        """i = identifier; h = hash"""
+        for record in records:
+            if self.extract_type == "dcatus":
+                i, h = record["identifier"], dataset_to_hash(sort_dataset(record))
+            if self.extract_type == "waf":
+                i, h = self.get_title_from_fgdc(record["content"]), dataset_to_hash(
+                    record["content"].decode(
+                        "utf-8"
+                    )  # TODO: is this the correct encoding?
+                )
+            self.records[i] = Record(
+                self,
+                i,
+                record,
+                h,
+            )
+
+    def ckan_to_id_hash(self, results: list[dict]) -> None:
+        for record in results:
+            self.ckan_records[record["identifier"]] = dataset_to_hash(
+                eval(
```

Comment:
why do we need eval here?

Reply:
this was my shortcut to convert a json str into a dict. i was running into situations where the input json str needed to be scrubbed of backslashes and single quotes. I recall trying to implement something like...
```python
record["dcat_metadata"].replace("\\","").replace("'",'"') 
```
but I still ran into a json decode error when using json.load( str ). I'll go back and get something working without eval